### PR TITLE
Add Timeout and TTL to LoadTest

### PR DESCRIPTION
**Pull request by wanlin31**
This commit adds Timeout and TTL to the LoadTest's schema and the
logics for controller to mark load test runs too long as error and
delete the LoadTest lives longer than TTL.

### CODE REVIEW

**Comment by codeblooded** [comment]
This is likely to be a merge conflict with #92, so we'll need to coordinate what goes in first.

**Review by codeblooded — CHANGES_REQUESTED** [review]
@wanlin31 There's a few thoughts I had so far. We can schedule a meeting to discuss some of them if you'd like. Overall, it's looking good 👍🏻

File: config/samples/go_example_loadtest.yaml
Comment (wanlin31) [reply config/samples/go_example_loadtest.yaml:57]: I was going to talk about this in our meeting :). I made TTL and Timeout as time.duration, unfortunately go 1.14 fixes the unit as ns, see (https://github.com/golang/go/issues/4712), and it doesn't take scientific notion when I played with it. 

Method I : Direct fill as duration 
Pro: fail fast 
Con: inconvenient to use, prone to human error, but it might be okay since we give templates

Method II : Direct fill as string, and parse to durations
Pro: you can have you own units, convenient to use
Con: format check depend on controller

Method III : Direct fill as int64, and convert to durations
Pro: more convenient than Method I to some extend
        fail fast

When I trying to explain the good and bad of three ways, I tend to agree that take the field as int64 could be the best for the scenario. I will change it. :)

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:152]: My thought is like this, open to better/other ways to do that definitely: we calculate the start time and the end time within status function, failing to update these field has impact on when do we re-queue the LoadTest for Timeout and TTL. 

Method I: We either delete the LoadTest and let the user know there is a problem and ask users to do it again.
Pro: We would not have orphaned Loadtest
con: we might delete the LoadTest necessarily.

Method II: ignore it 
Pro: We will not delete the LoadTest when the problem is only related to kubernetes.
Con:this LoadTest could in fact timeout but not mark as error, or live forever. 

I put it here just as one way, I want to learn about how Paulo and you think.

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:90]: I will update this.

File: status/status.go
Comment (wanlin31) [reply status/status.go:109]: I just need to have two flags to re-queue the LoadTest, best practice of doing this? 

Currently, timeout could be due to the test itself or the scheduling problem, adding TimeoutExceeded state tells you it is timed out, but you still don't know whether it is due to scheduling issue or the test itself, my feeling is this TimeoutExceeded state might not give a lot further information than error itself to users. I am not saying we should not add it, just some thoughts from a usage point of view.

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go]: I think about this again: 
If we delete the LoadTest failed to update at this moment, we could have delete the LoadTest unnecessarily only to save resource, it doesn't have to be this strict with saving resource. 

If we  just re queue it, the worst case scenario is that this LoadTest could live forever and its pods could be forever running and occupying the resource, if this happened a lot and the resource will be depleted, user interference will be necessary to clean up things like this, do not have to worry too much about this.

I convert it back to where it was like.

File: api/v1/loadtest_types.go
Comment (codeblooded) [inline api/v1/loadtest_types.go]: I'm not sure these should be mandatory. I think our system should have sane defaults. These should not be pointers if they are required. The pointers convey optionality ([API Conventions: Optional vs. Required](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required)).

Comment (wanlin31) [reply api/v1/loadtest_types.go]: I will fix the pointer type if we still want it mandatory, thank for you pointing out!

I think whether we want to have them mandatory or not is really depend on what we do if these values are not set. My understanding is we definitely need these values to perform garbage collection.  
1. If we are going to give them default values if not set, we could definitely make them optional, but we would need to implement them in cluster configuration and check if they are actually there. Moreover I think these two values are more of properties of load tests than the cluster. 
2. If we are going to reject the load test without these values in controller, we might just reject it before we get to scheduling.

**Comment by wanlin31** [comment]
I am still working on this Pr. Things I am working on:
1.  Just realized, even we made ttl and timeout mandatory, we still need to have checks to make sure they are reasonable values, and decide what to do if they are not.
2. I am moving the logics of calculating time for requeueAfter out, so that I could add test for it.
3. Thinking about if I could add better test for status.

File: api/v1/loadtest_types.go
Comment (codeblooded) [inline api/v1/loadtest_types.go:213]: I think there may be typos here, but otherwise good. 

```suggestion
	// TTL provide the longest time a LoadTest could live on the cluster.
```

File: config/samples/go_example_loadtest.yaml
Comment (codeblooded) [inline config/samples/go_example_loadtest.yaml:55]: I like that we add these to the examples, but I wish we could elaborate a little more on each one.

Maybe we could break them into each having their own comments and explain what a TTL or timeout does to effect a test.

As a developer, I feel like I may be confused to know what to put for both of them. It may be helpful to specify the timeout is the maximum amount of time a test is allowed to run. Meanwhile, the TTL is the amount of time the load test and its result status should live in memory before being garbage collected.

If I was using this framework for the first time, I probably would not be likely to check the docs for these fields. I'd probably just play around with these examples and read their comments.

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:76]: Great conversion.

nit: can we call this ttl and timeout or testTTL and testTimeout? We dropped the "load" from the test in a previous cleanup, because this is the LoadTestReconciler. The type of test is implied.

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:87]: nit: logs can be rendered as plain text or json. Let's make sure the fields, like the TTL all have the same case and name in all the log statements. I would suggest `ttl` or `testTTL`. JSON keys don't often start with caps.

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:94]: We can leave the name of the load test off, because it is automatically added to every log statement in reconcile. 

```suggestion
				log.Error(err, "fail to delete test")
```

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:166]: nit: let's keep the same case as the function. 

```suggestion
	requeueTime := getRequeueTime(previousStatus, test, log)
```

File: controllers/loadtest_controller_test.go
Comment (codeblooded) [inline controllers/loadtest_controller_test.go:768]: A reconciler is a large object, can we make this a pointer?

```suggestion
	var reconciler *LoadTestReconciler
```

File: status/status.go
Comment (codeblooded) [inline status/status.go:124]: Great! We should add a reason string and message explaining the timeout. You can see how I did this in #92, adding KubernetesError reason and a string message. 

The reason could be something like `TimeoutExceeded`.

**Review by codeblooded — CHANGES_REQUESTED** [review]
Looking very good. Very well done in my opinion. A few changes I'd like. 

I'm remarkably impressed by how much ginkgo and BDD tests come so naturally for you now.

File: config/samples/go_example_loadtest.yaml
Comment (codeblooded) [reply config/samples/go_example_loadtest.yaml:55]: We should also add them to other languages.

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:76]: I wanted to use ttl and timeout originally, but I found that in the same func we have a Timeout for reconciler (https://github.com/grpc/test-infra/blob/5c7cfabd47b59d9f7c35494be247b31414a2d431/controllers/loadtest_controller.go#L45), I will change them to testTTL and testTimeout, that is no problem :) Thank you for coming up names.

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:87]: I will check!

File: config/samples/go_example_loadtest.yaml
Comment (wanlin31) [reply config/samples/go_example_loadtest.yaml:55]: Yes, this just is an example for us to discuss if we want to do it the way it is now, once we are sure about this, I will add them to all other languages (go is the easiest/fastest to try). Thanks for reminding me!

For the explanation, I will separate them and explain more for each.

**Comment by wanlin31** [comment]
I have updated the items we talked about yesterday. I think it is okay to be reviewed.

File: config/samples/cxx_example_loadtest.yaml
Comment (codeblooded) [inline config/samples/cxx_example_loadtest.yaml:55]: nit: I think the state is past tense, Errored. I think this explains the details. I think @paulosjca may have feedback on wording.

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:158]: nit: I would keep consistent case with function and make the variable `requeueTime`.

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:168]:
```suggestion
// its current status is because the ttl and timeout are saved in its spec.
```

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:172]: nit: I think @paulosjca has reviewed me on spelling out small numbers before. He may want this to be changed from 0 to zero.

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:176]: What would happen in the following case? The controller sets the start time and requeues, but then the controller crashes and restarts. The queue is lost. How would this logic be triggered again? I have no idea how we'd handle that case.

File: controllers/loadtest_controller_test.go
Comment (codeblooded) [inline controllers/loadtest_controller_test.go:824]: It blocks should have different names. This allows us to see what failed quickly. So perhaps these should be more specific, like "returns 0 duration when ...".

File: controllers/loadtest_controller_test.go
Comment (codeblooded) [inline controllers/loadtest_controller_test.go:770]: Great job with tests. You are a native ginkgo/BDD speaker now.

**Review by codeblooded — COMMENTED** [review]
I think this is nearing ideal state. I have a few minor comments and a question that I don't have the answer to. It may be sufficient to not address that question for the moment and follow-up with an enhancement to address it.

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:172]: I remembered that :) Will fix them.

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:176]: I feel there is not any convenient way to deal with re queue of the event.

The worst case scenario:
1. Not re queue with timeout --> not able to catch it when it did went over time -->  the cleanup agent won't know about this --> would not stop the workers if they are still running --> will depend on TTL to delete them.
This worse case scenario would cause some resource waste, but if it happened too frequently, user interference might be required to resolved it. 

2. Not re queue to see it at TTL --> the load test will live on the cluster possibly forever --> possibility causing resource waste, this again would require human attention if this happened a lot. 

I agree it would be nice to have something that independent,  just do a cleanup up for everything on the cluster (delete the ones live out of TTL), I also agree this is more like a improvement later, if we do need this, it should not be too difficult to add down the road.

File: controllers/loadtest_controller_test.go
Comment (wanlin31) [reply controllers/loadtest_controller_test.go:824]: Right, I thought the conditions has to be in the context...I will modify it.

File: config/samples/cxx_example_loadtest.yaml
Comment (wanlin31) [reply config/samples/cxx_example_loadtest.yaml:55]: I definitely should change this since it has to be in agreement with states names in Loadtest.LoadTestState.

File: controllers/loadtest_controller.go
Comment (codeblooded) [inline controllers/loadtest_controller.go:158]: I hope this doesn't sound difficult, but I prefer the `requeueTime` and `getRequeueTime` casing, since it matches `ctrl.Result{Requeue: true}` and `ctrl.Result{RequeueAfter: ...}`.

**Review by codeblooded — APPROVED** [review]
LGTM after 1 fix and @paulosjca may have comments on wording for documentation. 

Let's hold off on merging until after #92 and a deployment, since they most certainly will conflict.

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:158]: No, you are helping find things I missed, I appreciate that! I fix it after the meeting.:) I also like to keep it tidy.

File: config/samples/cxx_example_loadtest.yaml
Comment (paulosjca) [inline config/samples/cxx_example_loadtest.yaml:53]: I would rewrite this as follows:

```yaml
  # timeoutSeconds is an integer field that indicates the longest time a test
  # is allowed to run, in seconds. Tests that run longer than the given value
  # will be marked as Errored and will no longer be allocated resources to run.
  # For example: timeoutSeconds: 900 indicates the timeout of this test
  # is 15min. The minimum valid value for this field is 1.
```

File: config/samples/cxx_example_loadtest.yaml
Comment (paulosjca) [inline config/samples/cxx_example_loadtest.yaml:61]: I would rewrite this as follows:

```yaml
  # ttlSeconds is an integer field that indicates how long a test is allowed to
  # live on the cluster, in seconds. Tests that live longer than the given value
  # will be deleted. For example: ttlSeconds: 86400 indicates the time-to-live
  # of this test is 24h. The minimum valid value for this field is 1.
```

File: config/samples/go_example_loadtest.yaml
Comment (paulosjca) [inline config/samples/go_example_loadtest.yaml:53]: The C++ comments apply to the other languages as well.

File: controllers/loadtest_controller.go
Comment (paulosjca) [inline controllers/loadtest_controller.go:222]: I would rewrite this as follows:

```go
// getRequeueTime takes a LoadTest and its previous status, compares the
// previous status of the load test with its updated status, and returns a
// calculated requeue time. If the test has just been assigned a start time
// (i.e., it has just started), the requeue time is set to the timeout value
// specified in the LoadTest. If the test has just been assigned a stop time
// (i.e., it has just terminated), the requeue time is set to the time-to-live
// specified in the LoadTest, minus its actual running time. In other cases,
// the requeue time is set to zero.
```

File: config/samples/cxx_example_loadtest.yaml
Comment (wanlin31) [reply config/samples/cxx_example_loadtest.yaml:53]: This is much more clearer! Thank you!

File: controllers/loadtest_controller.go
Comment (wanlin31) [reply controllers/loadtest_controller.go:222]: This is great, thank you for reviewing! I forgot to describe what the default value is. I would describe what the function does in all the cases next time, this is great to learn.

File: api/v1/loadtest_types.go
Comment (paulosjca) [inline api/v1/loadtest_types.go:269]: There is a typo here. I thought TimeoutExceeded was fine, but it is ok to change to TimeoutErrored as well.

Comment (wanlin31) [reply api/v1/loadtest_types.go]: Right, I think it is a type of error reason, so it made scene to have the 'error' word in the reason itself.

**Review by paulosjca — APPROVED** [review]

**Pull request merged**
